src/main/java/org/jenkinsci/plugins/workflow/log/BufferedBuildListener.java

```diff
@@ -0,0 +1,178 @@
+        private static final long MIN_RECURRENCE_PERIOD = 250; // ¼s
+        private static final long MAX_RECURRENCE_PERIOD = 10_000; // 10s
+        private static final float RECURRENCE_PERIOD_BACKOFF = 1.05f;
+
+        private long recurrencePeriod = MIN_RECURRENCE_PERIOD;
+
+        DelayBufferedOutputStream(OutputStream out) {
+            super(new FlushProofOutputStream(out)); // default buffer size: 8Kib
+            reschedule();
+        }
+
+        private void reschedule() {
+            Timer.get().schedule(new Flush(this), recurrencePeriod, TimeUnit.MILLISECONDS);
+            recurrencePeriod = Math.min((long) (recurrencePeriod * RECURRENCE_PERIOD_BACKOFF), MAX_RECURRENCE_PERIOD);
+        }
+
+        /** We can only call {@link BufferedOutputStream#flushBuffer} via {@link #flush}, but we do not wish to flush the underlying stream, only write out the buffer. */
+        private void flushBuffer() throws IOException {
+            ThreadLocal<Boolean> fauxFlushing = ((FlushProofOutputStream) out).fauxFlushing;
+            boolean orig = fauxFlushing.get();
+            fauxFlushing.set(true);
+            try {
+                flush();
+            } finally {
+                fauxFlushing.set(orig);
+            }
+        }
+        
+        @Override public void close() throws IOException {
+            // Ignored. We do not allow the stream to be closed from the remote side.
+        }
+        
+        void run() {
+            try {
+                flushBuffer();
+            } catch (IOException x) {
+                LOGGER.log(Level.FINE, null, x);
+            }
+            reschedule();
+        }
+        
+    }
+    
+    private static final class Flush implements Runnable {
+
+        /** Since there is no explicit close event, just keep flushing periodically until the stream is collected. */
+        private final Reference<DelayBufferedOutputStream> osr;
+        
+        Flush(DelayBufferedOutputStream os) {
+            osr = new WeakReference<>(os);
+        }
+        
+        @Override public void run() {
+            DelayBufferedOutputStream os = osr.get();
+            if (os != null) {
+                os.run();
+            }
+        }
+        
+    }
+
+    private static final class FlushProofOutputStream extends FilterOutputStream {
+
+        private final ThreadLocal<Boolean> fauxFlushing = new ThreadLocal<Boolean>() {
+            @Override protected Boolean initialValue() {
+                return false;
+            }
+        };
+
+        FlushProofOutputStream(OutputStream out) {
+            super(out);
+        }
+
+        @Override public void write(byte[] b, int off, int len) throws IOException {
+            out.write(b, off, len); // super method writes one byte at a time!
+        }
+        
+        @Override public void flush() throws IOException {
+            if (!fauxFlushing.get()) {
```

Comment:
nit: I find the naming kind of confusing here. This looks to me like a `FlushUnlessFlushingIsTemporarilyDisabledOutputStream` :), and I think `fauxFlushing` could be more clearly named `allowFlushing` and its initial value and uses could be inverted.

Reply:
Ack, can work on the naming a bit.